parcels/particlesetaos.py

```diff
@@ -0,0 +1,769 @@
+                 object that defines custom particle
+        :param lon: List of initial longitude values for particles
+        :param lat: List of initial latitude values for particles
+        :param depth: Optional list of initial depth values for particles. Default is 0m
+        :param time: Optional list of start time values for particles. Default is fieldset.U.time[0]
+        :param repeatdt: Optional interval (in seconds) on which to repeat the release of the ParticleSet
+        :param lonlatdepth_dtype: Floating precision for lon, lat, depth particle coordinates.
+               It is either np.float32 or np.float64. Default is np.float32 if fieldset.U.interp_method is 'linear'
+               and np.float64 if the interpolation method is 'cgrid_velocity'
+        Other Variables can be initialised using further arguments (e.g. v=... for a Variable named 'v')
+       """
+        return cls(fieldset=fieldset, pclass=pclass, lon=lon, lat=lat, depth=depth, time=time, repeatdt=repeatdt, lonlatdepth_dtype=lonlatdepth_dtype, **kwargs)
+
+    @classmethod
+    def from_line(cls, fieldset, pclass, start, finish, size, depth=None, time=None, repeatdt=None, lonlatdepth_dtype=None):
+        """Initialise the ParticleSet from start/finish coordinates with equidistant spacing
+        Note that this method uses simple numpy.linspace calls and does not take into account
+        great circles, so may not be a exact on a globe
+
+        :param fieldset: :mod:`parcels.fieldset.FieldSet` object from which to sample velocity
+        :param pclass: mod:`parcels.particle.JITParticle` or :mod:`parcels.particle.ScipyParticle`
+                 object that defines custom particle
+        :param start: Starting point for initialisation of particles on a straight line.
+        :param finish: End point for initialisation of particles on a straight line.
+        :param size: Initial size of particle set
+        :param depth: Optional list of initial depth values for particles. Default is 0m
+        :param time: Optional start time value for particles. Default is fieldset.U.time[0]
+        :param repeatdt: Optional interval (in seconds) on which to repeat the release of the ParticleSet
+        :param lonlatdepth_dtype: Floating precision for lon, lat, depth particle coordinates.
+               It is either np.float32 or np.float64. Default is np.float32 if fieldset.U.interp_method is 'linear'
+               and np.float64 if the interpolation method is 'cgrid_velocity'
+        """
+
+        lon = np.linspace(start[0], finish[0], size)
+        lat = np.linspace(start[1], finish[1], size)
+        if type(depth) in [int, float]:
+            depth = [depth] * size
+        return cls(fieldset=fieldset, pclass=pclass, lon=lon, lat=lat, depth=depth, time=time, repeatdt=repeatdt, lonlatdepth_dtype=lonlatdepth_dtype)
+
+    @classmethod
+    def monte_carlo_sample(cls, start_field, size, mode='monte_carlo'):
```

Comment:
well, I look into it again, but they needs each at least a minimum referral, because it's a class-method.
The case of the monte-carlo sample: that is structure-specific, hence that stays where it is.

Reply:
So, I had a look again into it. There are only 2 `from_<...>` methods that remain in the specialised classes (AOS, SOA, ...), which are 
`from_field(...)` and `from_particefile(...)`, in addition to `monte_carlo_sample(...)`.

`from_particlefile(...)` and `monte_carlo_sample(...)` are structure-dependent. `monte_carlo_sample(...)` is structure-dependent because iterating over the field and collecting the particles fast is done in SOA by iterate-over-variables and then LINQ-like selections, whereas for AOS and lists and so forth, it's done by iterate-over-particles followed by iterate-over-variables. `from_particlefile(...)` is structure-dependent because iteration and storage of data is particlefile-dependent - in AOS and SOA, we assume that `id == index` and thus use the stored ID's as ordering element in iterations. In lists, IDs and indices of read particles are different, hence there we (will) iterate over `indices` for particle set initialisation.

for `from_field(...)`, it's more of a question for me how the `cls` parameter in class methods are interpreted. Explanation:
we have a class method `from_field(...)` and a class method `monte_carlo_sample(...)` in `BaseParticleSet`. Assume the `from_field(...)` always does the same thing and looks like this:

```
    @classmethod
    def from_field(cls, fieldset, pclass, start_field, size, mode='monte_carlo', depth=None, time=None, repeatdt=None, lonlatdepth_dtype=None):
        lon, lat = cls.monte_carlo_sample(start_field, size, mode)
        return cls(fieldset=fieldset, pclass=pclass, lon=lon, lat=lat, depth=depth, time=time,
                   lonlatdepth_dtype=lonlatdepth_dtype, repeatdt=repeatdt)
```

in `BaseParticleSet::monte_carlo_sample(...)`, the function is (a) abstract and (b) only executes `pass`. Now, we have also the specialised implementations in `ParticleSetSOA::monte_carlo_sample(...)` and `ParticleSet:AOS:monte_carlo_sample(...)`, which do the actual particle sample, in the efficient way according to their structure. Now, if I create a ParticleSetAOS from a file, such as

```
pset = ParticleSetAOS.from_file(...)
```

, then the class method in `BaseParticleSet` is called. Now, in that instance, what is the `cls` parameter ? Is it `ParticleSetAOS` ? Perhaps yes, but then there is no explicit definition of `ParticleSetAOS::from_file(...)` - there is just one in the superclass `BaseParticleSet::from_file(...)`. So, logically, initially `cls = ParticleSetAOS`, but is then internally cast for the superclass call to `cls = BaseParticleSet`. Now if `BaseParticleSet::from_file(...)` calls `cls.monte_carlo_sample`, logic dictated that the function that is called is `BaseParticleSet::monte_carlo_sample(...)` - which is (for our intends and purposes) empty - that's a failure. What we want is that `ParticleSetAOS::monte_carlo_sample` is called. Thus, in order to achieve that, we would need to define `ParticleSetAOS::from_file(...)` (and a `from_file(...)` function  for all other specialised classes), so that `cls` always has the correct value. Thus, we need to replicate the `from_field(...)` function for all specialised classes. That's what logic dictates. Of course, if in this example, the value of `cls = ParticleSetAOS`, also if the function called is `BaseParticleSet::from_file(...)`, then my assumption is wrong and we could spare ourselves a repeated definition of `from_file(...)` in all derived classes. I am at a loss here cause there is nothing like a `class method` in C/C++ - we only have static functions that are completely class-independent, or class member functions where the `self` parameter is strictly defined by the specific class type. Thus I am not 100% sure how the `cls` parameter is interpreted in a class inheritance hierarchy in Python. Perhaps someone can comment here about it.

The other `from_<...>(...)` functions are in `BaseParticleSet` as they operate independently from a specific implementation.

Reply:
I think this little piece of code might provide a bit of insight, if you run it:
```py
class A(object):
    @classmethod
    def foo(cls):
        print(cls)

    @classmethod
    def oop(cls):
        print(cls)


class B(A):
    @classmethod
    def foo(cls):
        super().foo()

    @classmethod
    def bar(cls):
        cls.foo()


if __name__ == "__main__":
    B.bar()
    B.foo()
    B.oop()
```
Both will print `B` as the value of `cls`. I believe `cls` keeps the value of the object class that was instantiated / where the method was called on, and does not take its superclass if the method happens to be defined there. But you can play around a bit with this to replicate the situation in the from_... functions.

Reply:
indeed you're right